### PR TITLE
[Form][OptionsResolver] Show deprecated options definition on debug:form command

### PR DESCRIPTION
**Pull request by yceruto**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Next move after https://github.com/symfony/symfony/pull/27277.

It will look like this:

Use `--show-deprecated` option to show form types with deprecated options (example):
debug_types_with_deprecated_options

Use `--show-deprecated` option to show deprecated options of the given form type (example):
debug_deprecated_options_from_type

Deprecated option (example):
debug_deprecated_option_text

debug_deprecated_option_json

### CODE REVIEW

**Comment by iltar** [comment]
Will it be possible to use this command to find _all_ form types that have a deprecated option, and see which of those options are deprecated per type? That would make fixing them a walk in the park!

**Comment by yceruto** [comment]
@iltar sure, I'll add an option `--deprecated` to filter the result on `debug:form` and `debug:form <type>`.

Status: Needs Work

**Comment by yceruto** [comment]
Any update about this? 

@ogizanagi you're working on this command in the past, so will be great to have your opinion on this! thanks. :)

File: src/Symfony/Component/Form/Console/Descriptor/Descriptor.php [hunk 1]
```diff
@@ -128,9 +136,57 @@ protected function getOptionDefinition(OptionsResolver $optionsResolver, $option
...
         }
 
+        if (isset($definition['deprecationMessage']) && \is_string($definition['deprecationMessage'])) {
+            $definition['deprecationMessage'] = strtr($definition['deprecationMessage'], array('%name%' => $option));
+        }
+
         return $definition;
     }
 
+    protected function filterTypesByDeprecated(array $types): array
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/Console/Descriptor/Descriptor.php]: To be consistent with current code, should this rather live in the command and pass deprecated types & options to the descriptor as `$options`, so the descriptors' responsibility sticks to handling output?

File: src/Symfony/Component/Form/Console/Descriptor/TextDescriptor.php [hunk 2]
```diff
@@ -124,6 +123,31 @@ protected function describeOption(OptionsResolver $optionsResolver, array $optio
...
     }
 
+    private function buildTableRows(array $headers, array $options): array
+    {
+        $tableRows = array();
+        $count = \count(max($options));
+        for ($i = 0; $i < $count; ++$i) {
+            $cells = array();
+            foreach (array_keys($headers) as $group) {
+                if (isset($options[$group][$i])) {
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/Console/Descriptor/TextDescriptor.php]: Could be simplified a bit I guess:

```diff
diff --git a/src/Symfony/Component/Form/Console/Descriptor/TextDescriptor.php b/src/Symfony/Component/Form/Console/Descriptor/TextDescriptor.php
index 1e63aa0c68..3702d2f6b1 100644
--- a/src/Symfony/Component/Form/Console/Descriptor/TextDescriptor.php
+++ b/src/Symfony/Component/Form/Console/Descriptor/TextDescriptor.php
@@ -130,17 +130,12 @@ class TextDescriptor extends Descriptor
         for ($i = 0; $i < $count; ++$i) {
             $cells = array();
             foreach (array_keys($headers) as $group) {
-                if (isset($options[$group][$i])) {
-                    $option = $options[$group][$i];
-
-                    if (\is_string($option) && \in_array($option, $this->requiredOptions, true)) {
-                        $option .= ' <info>(required)</info>';
-                    }
-
-                    $cells[] = $option;
-                } else {
-                    $cells[] = null;
+                $option = $options[$group][$i] ?? null;
+                if ($option && \is_string($option) && \in_array($option, $this->requiredOptions, true)) {
+                    $option .= ' <info>(required)</info>';
                 }
+
+                $cells[] = $option;
             }
             $tableRows[] = $cells;
         }
```

File: src/Symfony/Component/Form/Tests/Fixtures/Descriptor/types_with_deprecated_options.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+    "builtin_form_types": [],
+    "service_form_types": [
+        "Symfony\\Component\\Form\\Tests\\Console\\Descriptor\\FooType"
+    ]
+}
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/Tests/Fixtures/Descriptor/types_with_deprecated_options.json]: Missing new line

**Review by ogizanagi — COMMENTED** [review]
Makes sense. Thanks @yceruto for working on this!

File: src/Symfony/Component/Form/Console/Descriptor/Descriptor.php (hunk 1 above)
Comment (yceruto) [reply src/Symfony/Component/Form/Console/Descriptor/Descriptor.php]: I agree, but in that case should we to move `collectOptions()`, `getParentOptionsResolver()`, etc, too?

**Comment by yceruto** [comment]
Thanks for the review @ogizanagi! I've fixed most comments.

File: src/Symfony/Component/Form/Console/Descriptor/Descriptor.php (hunk 1 above)
Comment (ogizanagi) [reply src/Symfony/Component/Form/Console/Descriptor/Descriptor.php]: I guess it's okay as we aim to describe OptionResolver and ResolvedFormTypeInterface instances. What is a bit weird here to me is injecting the form registry.
But no strong opinion.

Comment (yceruto) [reply src/Symfony/Component/Form/Console/Descriptor/Descriptor.php]: Okey, done.

File: src/Symfony/Component/Form/Command/DebugCommand.php [hunk 4]
```diff
@@ -58,6 +58,7 @@ protected function configure()
             ->setDefinition(array(
                 new InputArgument('class', InputArgument::OPTIONAL, 'The form type class'),
                 new InputArgument('option', InputArgument::OPTIONAL, 'The form type option'),
+                new InputOption('show-deprecated', null, InputOption::VALUE_NONE, 'Used to show deprecated options in form types'),
```
Comment (fabpot) [inline src/Symfony/Component/Form/Command/DebugCommand.php]: Display deprecated options in form types

File: src/Symfony/Component/Form/Command/DebugCommand.php [hunk 5]
```diff
@@ -75,6 +76,11 @@ protected function configure()
 
   <info>php %command.full_name% ChoiceType choice_value</info>
 
+Use the <info>--show-deprecated</info> option to display form types with deprecated options or the deprecated options of the given form type:
```
Comment (fabpot) [inline src/Symfony/Component/Form/Command/DebugCommand.php]: Should be cut at around 80 chars

**Review by fabpot — APPROVED** [review]
with minor comments

**Comment by yceruto** [comment]
@fabpot comments addressed, thanks! (AppVeyor failure is unrelated)

**Comment by fabpot** [comment]
Thank you @yceruto.

**Pull request merged**